SO-3382: SNOMED CT Query Language API

AAAlinaaa opened the pull request.

AAAlinaaa committed:
SO-3382: Implement Snomed query language

AAAlinaaa committed:
SO-3382: Add server side support to snomed query language

AAAlinaaa committed:
SO-3382: Update license headers, add pom.xml

AAAlinaaa committed:
Merge branch '6.x' into feature/SO-3382-implement_snomed_query_language

cmark commented on snomed/com.b2international.snowowl.snomed.api.rest/src/com/b2international/snowowl/snomed/api/rest/SnomedConceptRestService.java:
```diff
@@ -111,6 +111,10 @@
 			@RequestParam(value="ecl", required=false) 
 			final String eclFilter,
 			
+			@ApiParam(value="The Snomed query expression to match")
```
Value should be `SNOMED CT Query expression to match`.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/DefaultQlParser.java:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2018 B2i Healthcare Pte Ltd, http://b2i.sg
```
Copyright year should be `2019`.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/DefaultQlParser.java:
```diff
@@ -0,0 +1,94 @@
+/**
+ * @since 6.12
+ */
+public class DefaultQlParser implements QlParser {
```
I suggest renaming the Query Language services, APIs to use `SnomedQuery` instead of just `Ql`.
Couple of examples:
- `QlParser` -> `SnomedQueryParser`
- `DefaultQlParser` -> `DefaultSnomedQueryParser`
- `QlExpression` -> `SnomedQueryExpression`

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/QlSerializer.java:
```diff
@@ -0,0 +1,33 @@
+public interface QlSerializer {
+
+	/**
+	 * Serializes a {@link Domainmodel} to its brief syntax equivalent.
```
Fix javadoc reference to invalid `DomainModel` type.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/SnomedQlEvaluationRequest.java:
```diff
@@ -0,0 +1,149 @@
+
+	private static final long serialVersionUID = 8932162693072727864L;
+
+	private final PolymorphicDispatcher<Promise<Expression>> dispatcher = PolymorphicDispatcher.createForSingleTarget("eval", 2, 2, this);
```
Make this variable transient and add `JsonIgnore` annotation as well, so it won't get serialized through the network.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/SnomedQlEvaluationRequest.java:
```diff
@@ -0,0 +1,149 @@
+	protected Promise<Expression> eval(BranchContext context, final TermFilter termFilter) {
+		final String term = termFilter.getTerm();
+		
+		return SnomedRequests.prepareSearchDescription()
```
Missing limit setting, this will only return the first `50` term matches.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/SnomedQlEvaluationRequest.java:
```diff
@@ -0,0 +1,149 @@
+	
+	protected Promise<Expression> eval(BranchContext context, final Disjunction disjunction) {
+		return Promise.all(evaluate(context, disjunction.getLeft()), evaluate(context, disjunction.getRight()))
+				.then(new Function<List<Object>, Expression>() {
```
Use lambdas, please.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/SnomedQlEvaluationRequestBuilder.java:
```diff
@@ -0,0 +1,43 @@
+		extends BaseRequestBuilder<SnomedQlEvaluationRequestBuilder, BranchContext, Promise<Expression>> 
+		implements RevisionIndexRequestBuilder<Promise<Expression>> {
+
+	private final SnomedQlEvaluationRequest req = new SnomedQlEvaluationRequest();
```
This request should not be constructed until you actually need to do it in `doBuild()`, please refactor.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptSearchRequest.java:
```diff
@@ -75,6 +76,11 @@
+		/**
+		 * Query language expression to match
+		 */
+		QL,
```
Rename this OptionKey literal to `QUERY` and fix javadoc.

cmark commented on snomed/com.b2international.snowowl.snomed.ql/.classpath:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Remove the `.antlr-generator-3.2.0-patch.jar` from the `snomed.ql` module and use a `.gitignore` file to prevent it from being added to git later.

cmark commented on snomed/com.b2international.snowowl.snomed.ql/src/com/b2international/snowowl/snomed/QL.xtext:
```diff
@@ -0,0 +1,51 @@
+Query:
+	{Query} (disjunction=Disjunction)?;
+	
+Disjunction:
```
Supporting `OR`/`AND`/`MINUS` operators require handling of ambiguity issues, when you specify expressions like this: `A OR B AND C MINUS D`.
Please fix it by introducing recursion in the language via nested expressions.

cmark commented on snomed/com.b2international.snowowl.snomed.ql/src/com/b2international/snowowl/snomed/generator/QLGenerator.xtend:
```diff
@@ -0,0 +1,37 @@
+class QLGenerator extends AbstractGenerator {
+
+	override void doGenerate(Resource resource, IFileSystemAccess2 fsa, IGeneratorContext context) {
+//		fsa.generateFile('greetings.txt', 'People to greet: ' + 
```
Remove unnecessary generated code.

cmark commented on snomed/com.b2international.snowowl.snomed.ql/src/com/b2international/snowowl/snomed/validation/QLValidator.xtend:
```diff
@@ -0,0 +1,37 @@
+ */
+class QLValidator extends AbstractQLValidator {
+	
+//	public static val INVALID_NAME = 'invalidName'
```
Add ambiguity validation here.

AAAlinaaa committed:
SO-3382: Update headers, rename QL services, remove unnecessary ...

...generated code, .jar file, add missing limit from snomed request,
fix outdated references, refactor snomed query request builder

AAAlinaaa committed:
SO-3382: Update query language model to include nested expressions

AAAlinaaa committed:
SO-3382 Refactor packages, add ambiguity validation

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/SnomedQueryEvaluationRequest.java:
```diff
@@ -0,0 +1,155 @@
+		
+		return SnomedRequests.prepareSearchDescription()
+				.all()
+				.filterByTerm(term)
```
Add active filter and return only the `ID` and `CONCEPT_ID` fields.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ql/SnomedQueryExpression.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public final class SnomedQueryExpression {
+
+	private final String ql;
```
This field still named `ql`, please rename.

cmark commented on snomed/com.b2international.snowowl.snomed.model/src/com/b2international/snowowl/snomed/SnomedConstants.java:
```diff
@@ -386,6 +386,11 @@ private Concepts() { }
 		public static final String SEARCH_TERM = "9221000132108";
 		public static final String SHORT_NAME = "9211000132103";
+		public static final String SIXTY_CHAR_SHORT_NAME = "9251000132104";
+		public static final String INGREDIENT_SYNONYM = "5981000132100";
```
⚠️ SG IDs in `SnomedConstants`! Move them to `SddConcepts`.

cmark commented on snomed/com.b2international.snowowl.snomed.ql/plugin.xml:
```diff
@@ -0,0 +1,10 @@
+<plugin>
+	<extension point="org.eclipse.emf.ecore.generated_package">
+		<package 
+			uri = "http://www.b2international.com/snowowl/snomed/Ql"
```
Make sure this entry is properly set. Check the `plugin.xml_gen` file.

cmark commented on snomed/com.b2international.snowowl.snomed.ql/plugin.xml_gen:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Remove this file from Git and add it to `.gitignore`.

cmark reviewed (approved):
Please add a `.gitignore` file to the root of the `snomed.ql` project.

AAAlinaaa committed:
SO-3382: Rename variable, remove sdd constants from snomed constants...

..., add .gitignore, add filter for active concepts only in query
evaluation and expand, description and concept ids, delete .xml_gen
file, capitalize package uri plugin.xml

AAAlinaaa committed:
SO-3382: Rename query variable on REST api

AAAlinaaa committed:
SO-3382: Remove plugin.xml_gen

cmark commented on snomed/com.b2international.snowowl.snomed.ql/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,35 @@
+Manifest-Version: 1.0
```
This project is still missing:
- a `pom.xml` file and `module` entry in the corresponding aggregator pom
- a feature.xml entry in `snomed.core.feature`

AAAlinaaa committed:
SO-3382: Update .xtendbins

AAAlinaaa committed:
SO-3382: Include project in aggregator pom.xml, and feature.xml

cmark reviewed (approved):
LGTM!
